[hal/vk] Rework Submission and Surface Synchronization

cwfitzgerald opened the pull request:
**Connections**

* Closes #5559
* Closes #5693 
* Maybe #4919
* Maybe #5374
* Maybe #5491
* Probably not #3452.

**Description**

As described in #5559, our submission and surface synchronization was totally messed up. I've revamped it, made smarter classes to help keep all the bookkeeping straight, and generally make the code cleaner and easier to understand.

**Testing**

works on my machine lel

cwfitzgerald committed:
[hal/vk] Rework Submission and Surface Synchronization

cwfitzgerald committed:
Destory

cwfitzgerald committed:
Comments

JMS55 commented:
Works fine for me running the cube example with VK validation layers enabled
```
[2024-05-09T06:03:33Z INFO  wgpu_core::instance] Adapter Vulkan AdapterInfo { name: "Intel(R) Arc(TM) Graphics", vendor: 32902, device: 32085, device_type: IntegratedGpu, driver: "Intel Corporation", driver_info: "101.5382", backend: Vulkan }
```

Vecvec commented:
This fixes an annoying access violation termination (that I thought was a driver bug) on program exit after a present.

hecrj commented:
Unfortunately, it does not fix #5644 for me. Same constant validation error on presentation:

```
[2024-05-09T08:16:24Z ERROR wgpu_hal::vulkan::instance] VALIDATION [VUID-vkAcquireNextImageKHR-semaphore-01779 (0x5717e75b)]
        Validation Error: [ VUID-vkAcquireNextImageKHR-semaphore-01779 ] Object 0: handle = 0xba7514000000002a, type = VK_OBJECT_TYPE_SEMAPHORE; | MessageID = 0x5717e75b | vkAcquireNextImageKHR():  Semaphore must not have any pending operations. The Vulkan spec states: If semaphore is not VK_NULL_HANDLE it must not have any uncompleted signal or wait operations pending (https://www.khronos.org/registry/vulkan/specs/1.3-extensions/html/vkspec.html#VUID-vkAcquireNextImageKHR-semaphore-01779)
[2024-05-09T08:16:24Z ERROR wgpu_hal::vulkan::instance]         objects: (type: SEMAPHORE, hndl: 0xba7514000000002a, name: ?)
```

The actual validation error only happens with a `debug` build, although `release` feels much slower than `0.19` as well. I believe something is very slow and the errors actually only show up in `debug` mode.

Flamegraph of `cube` seems to indicate this—with `present` taking a huge amount of time:

![flamegraph](https://github.com/gfx-rs/wgpu/assets/518289/c0e06cd3-25ad-456d-a7de-1e3888271805)

For comparison, same drivers with `0.19`:

![flamegraph](https://github.com/gfx-rs/wgpu/assets/518289/21e734b6-1d5c-4fc4-b4c6-47eac9dd34b7)

Even if I enable `Immediate` present mode, frames seem to take twice as much time to be rendered with `0.20`.

cwfitzgerald committed:
Nvidia

cwfitzgerald committed:
Novideo

cwfitzgerald committed:
Nits

cwfitzgerald commented:
@hecrj could you try again with these latest commits, someone else who can reproduce the validation error says that this fixed it.

cwfitzgerald committed:
Fix raw-gles

hecrj commented:
@cwfitzgerald Yep! The latest changes fix the validation errors. Thanks!

I'm still noticing a bit of worse performance in `debug` mode, but it doesn't seem to be there in `release` mode anymore. Flamegraphs look the same as before, but I imagine it's expected since now synchronization works properly and there may be more waiting. No errors in any case!

jimblandy commented on wgpu-hal/src/lib.rs:
```diff
@@ -486,9 +486,15 @@ pub trait Surface: WasmNotSendSync {
+    ///
+    /// # Safety
+    ///
+    /// - The fence must be the same fence passed to all [`Queue::submit`]s
+    ///   that used textures acquired from this surface.
```
This is unclear. I'm using textures acquired from this surface all the time, with many different submissions.

jimblandy replied:
This makes more sense to me now. Resolving.

jimblandy commented on wgpu-hal/src/vulkan/mod.rs:
```diff
@@ -149,6 +140,103 @@ pub struct Instance {
 
+/// The semaphores used to synchronize the swapchain image acquisition.
+///
+/// One of these is created per swapchain image and is used for the swapchain
```
This is unclear. A swapchain has several images. Each has its own `SwapchainSemaphores`. What does it mean for all of them to be "used for the swapchain"?

jimblandy commented:
I'm actually rewriting these comments as I review so if you're not attached to them then don't worry too much about fixing anything.

cwfitzgerald commented:
Yeah that's fine - I was just trying to get something there.

Wumpf commented on wgpu-hal/src/lib.rs:
```diff
@@ -675,7 +681,7 @@ pub trait Queue: WasmNotSendSync {
     /// Submit `command_buffers` for execution on GPU.
     ///
-    /// If `signal_fence` is `Some(fence, value)`, update `fence` to `value`
+    /// If `signal_fence` is `(fence, value)`, update `fence` to `value`
     /// when the operation is complete. See [`Fence`] for details.
```
```suggestion
    /// Will update `fence` to `value` when the operation is complete.
    /// See [`Fence`] for details.
```

jimblandy replied:
I've incorporated a change similar to this, so I'm going to resolve this.

jimblandy commented on wgpu-hal/src/vulkan/mod.rs:
```diff
@@ -359,18 +463,70 @@ pub struct Device {
 
+/// Semaphores that a given submission should wait on and signal.
+struct RelaySemaphoreState {
+    wait: Option<vk::Semaphore>,
+    signal: vk::Semaphore,
```
My reading is that you can actually just use one semaphore to order submits, for both the signal and wait operations.

The rules I can find about signaling and waiting just say they have to happen in the natural order, and a submission will definitely wait before signalling. So the semaphore will always be in the right state for the next operation.

It's not a problem just to enqueue multiple operations that will wait and signal the same semaphore - if it were, then you couldn't handle an arbitrary number of submissions just by swapping two semaphores back and forth. Certainly when you're doing that, half the submissions are waiting on the same semaphore. So if that's permitted, you might as well just use one semaphore for everything.

The text of `VUID-VkSubmitInfo2-semaphore-03881` explicitly mentions the possibility of semaphores appearing in both lists on the same submit:

> If the same semaphore is used as the semaphore member of both an element of `pSignalSemaphoreInfos` and `pWaitSemaphoreInfos`, and that semaphore is a timeline semaphore, the value member of the `pSignalSemaphoreInfos` element must be greater than the value member of the `pWaitSemaphoreInfos` element

That's describing timeline semaphores, but it shows that the authors are expecting this possibility. And there's no VUID saying that the same semaphore can't appear in both lists.

jimblandy replied:
As discussed in Matrix (my bad), technically what I said above is correct, but we want to keep the semaphore-swapping code for the time being because of https://gitlab.freedesktop.org/mesa/mesa/-/issues/5508 .

jimblandy commented:
@cwfitzgerald What do you think of this? https://github.com/gfx-rs/wgpu/commit/da543fb51

My hope was that this would make the states of `RelaySemaphores` a bit clearer to a new reader. If people should not use `wait` unless `should_wait` is true, then let's actually force people to check that condition before they can even get at the value: classic `Option`. But it does make `advance` fallible and require us to pass the device.

[edit: made this a little nicer still: `RelaySemaphoreState` is no longer necessary]

jimblandy commented:
@cwfitzgerald Here's an even more radical suggestion: https://github.com/gfx-rs/wgpu/commit/b5d52fed45e4f89213e6fe5870e99126865711aa

Just use a single semaphore for all queue ordering.

cwfitzgerald commented:
@jimblandy i think we can go for it, the only question is if we want to still avoid this deadlock in anv mentioned in the old code, it if we consider that old enough to ignore.    

```
/// It would be correct to use a single semaphore there, but 
/// [Intel hangs in `anv_queue_finish`](https://gitlab.freedesktop.org/mesa/mesa/-/issues/5508).
```

I should have perseved that comment somehow.

jimblandy commented:
@cwfitzgerald:

> I should have perseved that comment somehow.

Ahh, yes, I was reviewing by checking out the branch and just wandering around with rust-analyzer, so I didn't notice that the patch removed the comment. If that's the sole rationale for the entire semaphore-swapping doohickey then we had definitely better have the comment.

Kvark filed that bug in Oct 2021. That's not that long ago, so I think we'd better keep it. It sounds like a bear to debug, and I'd rather not have to debug it again.

jimblandy commented:
So I guess the remaining question is what you think of [this one](https://github.com/gfx-rs/wgpu/pull/5681#issuecomment-2118327977).

jimblandy committed:
[hal vulkan] Introduce `DeviceShared::new_binary_semaphore`.

Introduce a utility function for making binary semaphores, and use it
where appropriate.

jimblandy committed:
[hal vulkan] Turn `RelaySemaphores::wait` into an option.

This allows us to delete `should_wait`, and makes
`RelaySemaphoreState` the same as `RelaySemaphores`, so we can just
delete the former, and have `RelaySemaphores::advance` return a clone
of `self`'s current state.

jimblandy commented:
So here's the latest version of my suggestion:
https://github.com/jimblandy/wgpu/compare/hal-vk-wait-semaphore-is-option%7E2...jimblandy:wgpu:hal-vk-wait-semaphore-is-option

cwfitzgerald commented:
Looks fine to me, could you either PR to my fork or just push to it?

jimblandy commented:
I still want to get a handle on the swapchain semaphores. I've got some work-in-progress docs for those too. I'll take care of this tomorrow morning.

jimblandy commented on wgpu-hal/src/vulkan/mod.rs:
```diff
@@ -149,6 +140,99 @@ pub struct Instance {
+/// present operation. We need use one per submit as we don't know at submit time which
+/// submission will be the last to use the image, so we add semaphores to them all.
+#[derive(Debug)]
+struct SwapchainSemaphores {
```
Could we rename this `SwapchainImageSemaphores`, since it's per image, not per swapchain?

cwfitzgerald replied:
Sure

jimblandy commented on wgpu-hal/src/vulkan/mod.rs:
```diff
@@ -702,58 +871,85 @@ impl crate::Queue for Queue {
-        for &surface_texture in surface_textures {
-            wait_stage_masks.push(vk::PipelineStageFlags::TOP_OF_PIPE);
-            wait_semaphores.push(surface_texture.wait_semaphore);
+        // We lock access to all of the semaphores. This may block if two submissions are in flight at the same time.
```
Please check my understanding, but I think this can't actually block. All submissions using any given `SurfaceTexture` are required to use the same `Fence`, and as explained in the docs for `wgpu_hal::Queue::submit`, all submissions using the same `Fence` must be synchronized to occur in a particular order.

cwfitzgerald replied:
Ah yes, good point, then lets move these all to try_lock.

jimblandy commented on wgpu-hal/src/vulkan/instance.rs:
```diff
@@ -1011,12 +1016,28 @@ impl crate::Surface for super::Surface {
+            .try_lock()
+            .expect("Failed to lock a SwapchainSemaphores.");
+
+        // Wait for the previously acquired image used by the semaphore to be available.
+        swapchain.device.wait_for_fence(
+            fence,
+            locked_swapchain_semaphores.previously_used_submission_index,
+            timeout_ns,
+        )?;
```
What does this actually accomplish? Aren't we going to wait for the `acquire` semaphore anyway?

cwfitzgerald replied:
We must wait until the cpu timeline is sure that the semaphore has been signaled and reset. Otherwise there will be two acquire operations that could signal the same semaphore.

jimblandy replied:
So, in other words:

If we were to remove this wait, then acquisition, command submission, and presentation would all set stuff in motion in the presentation engine or on the GPU, but as far as the CPU is concerned, they'd all return without blocking, so the CPU could just run the process full speed ahead, blow through the entire swapchain, come around to the front of the ring buffer again, and make a fresh call to `vkAcquireNextImage` passing the same semaphore as last time. This wait is the only thing that rate-limits that process to match actual presentation speed.

Is that correct?

jimblandy replied:
I've pushed another commit documenting this. If my explanation is correct, please mark this as resolved.

cwfitzgerald replied:
> This wait is the only thing that rate-limits that process to match actual presentation speed.

Yes this is correct.

jimblandy committed:
[hal] Document `Surface` methods and some presentation constraints.

jimblandy committed:
[hal vulkan] Document `SwapchainSemaphores`.

jimblandy commented:
@cwfitzgerald I just pushed two doc commits. Could you look them over for accuracy?

jimblandy committed:
[hal vulkan] Rename `SwapchainSemaphores` -> `SwapchainImageSemaphores`

jimblandy commented on wgpu-hal/src/vulkan/mod.rs:
```diff
@@ -793,19 +1058,20 @@ impl crate::Queue for Queue {
         let ssc = swapchain.as_mut().unwrap();
+        let mut swapchain_semaphores = texture.surface_semaphores.lock();
+
+        // debug_assert_eq!(
+        //     Arc::as_ptr(&texture.surface_semaphores),
+        //     Arc::as_ptr(&ssc.surface_semaphores[ssc.next_semaphore_index]),
+        //     "Trying to use a surface texture that does not belong to the current swapchain."
+        // );
```
This should either be uncommented or removed, but we shouldn't leave it in limbo like this.

cwfitzgerald replied:
Lets remove it.

jimblandy committed:
Doc `SwapchainImageSemaphores::previously_used_submission_index`.

jimblandy committed:
Incorporate Wumpf's suggestions.

jimblandy commented:
rebased on current trunk

cwfitzgerald committed:
Apply suggestions
